File: forge-core/internal/capabilitygrantcontract/vocabulary.go
package capabilitygrantcontract

import (
	"fmt"
	"slices"
)

func validateVocabulary(vocabulary map[string]any) error {
	if err := requireKeys(vocabulary, "api_version", "canonicalization", "effects", "kind",
		"vocabulary_sha256"); err != nil {
		return fmt.Errorf("effect vocabulary: %w", err)
	}
	if err := requireStringLiteral(vocabulary, "api_version",
		"forgeos.governance.effect-vocabulary/v1"); err != nil {
		return err
	}
	if err := requireStringLiteral(vocabulary, "canonicalization", "forgeos.canonical-json/v1"); err != nil {
		return err
	}
	if err := requireStringLiteral(vocabulary, "kind", "EffectVocabulary"); err != nil {
		return err
	}
	effects, err := arrayValue(vocabulary, "effects")
	if err != nil || len(effects) != len(frozenEffects) {
		return fmt.Errorf("effects must contain the frozen %d descriptors", len(frozenEffects))
	}
	for index, expected := range frozenEffects {
		node, ok := effects[index].(map[string]any)
		if !ok {
			return fmt.Errorf("effect %d must be an object", index)
		}
		if err := validateDescriptor(node, expected); err != nil {
			return fmt.Errorf("effect %d: %w", index, err)
		}
	}
	if err := validateCanonicalByteLimit(vocabulary, maxVocabularyBytes, "effect vocabulary"); err != nil {
		return err
	}
	return validateVocabularyDigest(vocabulary)
}

func validateDescriptor(node map[string]any, expected effectDescriptor) error {
	if err := requireKeys(node, "allowed_scope_kinds", "effect_id", "production_restriction",
		"required_scope_kinds", "scope_profile"); err != nil {
		return err
	}
	checks := map[string]string{
		"effect_id": expected.id, "production_restriction": expected.restriction,
		"scope_profile": expected.profile,
	}
	for key, value := range checks {
		if err := requireStringLiteral(node, key, value); err != nil {
			return err
		}
	}
	allowed, err := readStringArray(node, "allowed_scope_kinds", 1, 9)
	if err != nil || !slices.Equal(allowed, expected.allowed) {
		return fmt.Errorf("allowed_scope_kinds differs from frozen vocabulary")
	}
	required, err := readStringArray(node, "required_scope_kinds", 1, 9)
	if err != nil || !slices.Equal(required, expected.required) {
		return fmt.Errorf("required_scope_kinds differs from frozen vocabulary")
	}
	return nil
}

func validateVocabularyDigest(vocabulary map[string]any) error {
	claimed, err := stringValue(vocabulary, "vocabulary_sha256")
	if err != nil || claimed != frozenVocabularySHA256 {
		return fmt.Errorf("vocabulary_sha256 is not the frozen v1 digest")
	}
	preimage := cloneNode(vocabulary)
	preimage["vocabulary_sha256"] = ""
	computed, err := digestNode(vocabularyDigestDomain, preimage)
	if err != nil || computed != claimed {
		return fmt.Errorf("vocabulary_sha256 does not match canonical vocabulary")
	}
	return nil
}

func readStringArray(parent map[string]any, key string, minimum, maximum int) ([]string, error) {
	values, err := arrayValue(parent, key)
	if err != nil || len(values) < minimum || len(values) > maximum {
		return nil, fmt.Errorf("%s item count must be %d..%d", key, minimum, maximum)
	}
	result := make([]string, len(values))
	for index, value := range values {
		text, ok := value.(string)
		if !ok || validateText(text, key, 160) != nil {
			return nil, fmt.Errorf("%s item %d must be bounded text", key, index)
		}
		result[index] = text
	}
	if !sortedUniqueStrings(result) {
		return nil, fmt.Errorf("%s must be sorted and unique", key)
	}
	return result, nil
}

func requireStringLiteral(parent map[string]any, key, expected string) error {
	value, err := stringValue(parent, key)
	if err != nil || value != expected {
		return fmt.Errorf("%s must equal %q", key, expected)
	}
	return nil
}

File: skills/policy-authority/scripts/_vendor/capability_grant_contract/vocabulary.py
"""Frozen 21-effect vocabulary validation."""

from __future__ import annotations

from typing import Any

from .canonical import ContractError, bounded_canonical_json, bounded_digest
from .constants import (CANONICALIZATION, EFFECTS, EFFECT_SPECS, VOCABULARY_API,
                        MAX_VOCABULARY_BYTES, VOCABULARY_DOMAIN, VOCABULARY_KIND,
                        VOCABULARY_SHA256)
from .shape import require_keys, sha256


def _expected_definition(effect_id: str) -> dict[str, Any]:
    allowed, required, restriction, profile = EFFECT_SPECS[effect_id]
    return {
        "allowed_scope_kinds": list(allowed),
        "effect_id": effect_id,
        "production_restriction": restriction,
        "required_scope_kinds": list(required),
        "scope_profile": profile,
    }


def vocabulary_sha256(value: dict[str, Any]) -> str:
    bounded_canonical_json(value, MAX_VOCABULARY_BYTES, "effect vocabulary")
    payload = dict(value)
    payload["vocabulary_sha256"] = ""
    return bounded_digest(VOCABULARY_DOMAIN, payload, MAX_VOCABULARY_BYTES,
                          "effect vocabulary")


def validate_vocabulary(value: Any) -> dict[str, Any]:
    keys = {"api_version", "canonicalization", "effects", "kind", "vocabulary_sha256"}
    node = require_keys(value, "effect vocabulary", keys)
    if node["api_version"] != VOCABULARY_API or node["kind"] != VOCABULARY_KIND:
        raise ContractError("effect vocabulary envelope is not v1")
    if node["canonicalization"] != CANONICALIZATION:
        raise ContractError("effect vocabulary canonicalization is unsupported")
    bounded_canonical_json(node, MAX_VOCABULARY_BYTES, "effect vocabulary")
    effects = node["effects"]
    if not isinstance(effects, list) or len(effects) != len(EFFECTS):
        raise ContractError("effect vocabulary must contain the frozen 21 definitions")
    for index, (definition, effect_id) in enumerate(zip(effects, EFFECTS)):
        if definition != _expected_definition(effect_id):
            raise ContractError(f"effect vocabulary definition {index} drifted from v1")
    sha256(node["vocabulary_sha256"], "vocabulary_sha256")
    actual = vocabulary_sha256(node)
    if actual != VOCABULARY_SHA256 or node["vocabulary_sha256"] != VOCABULARY_SHA256:
        raise ContractError("effect vocabulary is not the frozen v1 vocabulary")
    return node
